<img src="../assets/a_eyes_readme.gif" style="float:right ; margin: 10px ; width:300px;"> 


<h1><left>Prediction and Analysis of Degree of Suicidal
Ideation in Online Content</left></h1>
<h4><left>by- Rajswi Lochan(Roll no-1806095)</left></h4>
<h4><left> Prince Sinha(Roll no-1806132)</left></h4>
<h4><left> Parth Sharma(Roll no-1806172)</left></h4>

___

## 1. Data Collection
- For this project, we will be using Reddit's API to collect posts from two subreddits: "r/depression" and "r/SuicideWatch"
- We aim to automate as much of this process as possible into neat functions to enable repeatability on the data collection front.
- When collecting data from servers, we will create a randomized delay between requests as a consideration to Reddit's servers and security staff.

> Note: Data in this notebook was collected on 10th October 2020. Do note that if we run the code again on another day, it will result in a new set of posts being scraped.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import requests
import time
import pandas as pd
from random import randint

### 1.1 Exploring the HTML architecture of the r/depression subreddit page 

In [ ]:
#WE WILL SCRAPE THE r/depression AND r/SuicideWatch SUBREDDITS
#LET'S START BY EXPLORING THE HTML INNARDS OF THE FORMER
url_1 = "https://www.reddit.com/r/depression.json"

In [ ]:
#DEFINING A USER AGENT AND MAKING SURE STATUS IS GOOD TO GO
headers = {"User-agent" : "Sam He"}
res = requests.get(url_1, headers=headers)
res.status_code

200

In [ ]:
#PEEKING AT WHAT OUR DATA WILL LOOK LIKE
depress_json = res.json()
depress_json

{'data': {'after': 't3_jrxz5c',
  'before': None,
  'children': [{'data': {'all_awardings': [{'award_sub_type': 'GROUP',
       'award_type': 'global',
       'awardings_required_to_grant_benefits': 3,
       'coin_price': 300,
       'coin_reward': 250,
       'count': 1,
       'days_of_drip_extension': 0,
       'days_of_premium': 0,
       'description': 'THIS right here! Join together to give multiple This awards and see the award evolve in its display and shower benefits for the recipient. For every 3 This awards given to a post or comment, the author will get 250 coins.',
       'end_date': None,
       'giver_coin_reward': None,
       'icon_format': None,
       'icon_height': 2048,
       'icon_url': 'https://i.redd.it/award_images/t5_22cerq/vu6om0xnb7e41_This.png',
       'icon_width': 2048,
       'id': 'award_68ba1ee3-9baf-4252-be52-b808c1e8bdc4',
       'is_enabled': True,
       'is_new': False,
       'name': 'This',
       'penny_donate': None,
       'penny_price': No

In [ ]:
#THE REDDIT DATA SEEMS TO BE ORGANISED AS A DICTIONARY
#LET'S GET ITS KEYS
sorted(depress_json["data"].keys())

['after', 'before', 'children', 'dist', 'modhash']

In [ ]:
#WITH SOME HELP FROM THIS YOUTUBE TUTORIAL: https://www.youtube.com/watch?v=5Y3ZE26Ciuk
#WE FIND OUT THAT THE after KEY IS THE QUERY STRING THAT WILL...
#INDICATE IN OUR URL THAT WE WANT TO SEE THE NEXT 25 POSTS AFTER THE after "CODE"

depress_json["data"]["after"]

't3_jrxz5c'

In [ ]:
#DOUBLE CONFIRMING THAT THE PREVIOUS AFTER KEY IS REALLY THE LAST ITEM ON OUR PAGE
[post["data"]["name"] for post in depress_json["data"]["children"]]

['t3_doqwow',
 't3_iq10oq',
 't3_jrla14',
 't3_jrtbuz',
 't3_jrk7w9',
 't3_jrpu3j',
 't3_jr9klb',
 't3_jrx773',
 't3_jrmqkj',
 't3_jrwlk0',
 't3_jrpvjj',
 't3_jroq0h',
 't3_jrlsor',
 't3_jrvxck',
 't3_jrufwm',
 't3_jrnpc2',
 't3_jrl9tf',
 't3_jrmnu1',
 't3_jrois1',
 't3_jrwyau',
 't3_jrv6qs',
 't3_jrurvx',
 't3_jrwcrr',
 't3_jrx6j0',
 't3_jrwof9',
 't3_jri35b',
 't3_jrxz5c']

In [ ]:
#CHECKING OUT THE NUMBER OF POSTS IN ONE PAGE
len(depress_json["data"]["children"])

27

In [ ]:
# OH, WE CAN DATAFRAME IT. 
pd.DataFrame(depress_json["data"]["children"])

kind                                               data
0    t3  {'approved_at_utc': None, 'subreddit': 'depres...
1    t3  {'approved_at_utc': None, 'subreddit': 'depres...
2    t3  {'approved_at_utc': None, 'subreddit': 'depres...
3    t3  {'approved_at_utc': None, 'subreddit': 'depres...
4    t3  {'approved_at_utc': None, 'subreddit': 'depres...
5    t3  {'approved_at_utc': None, 'subreddit': 'depres...
6    t3  {'approved_at_utc': None, 'subreddit': 'depres...
7    t3  {'approved_at_utc': None, 'subreddit': 'depres...
8    t3  {'approved_at_utc': None, 'subreddit': 'depres...
9    t3  {'approved_at_utc': None, 'subreddit': 'depres...
10   t3  {'approved_at_utc': None, 'subreddit': 'depres...
11   t3  {'approved_at_utc': None, 'subreddit': 'depres...
12   t3  {'approved_at_utc': None, 'subreddit': 'depres...
13   t3  {'approved_at_utc': None, 'subreddit': 'depres...
14   t3  {'approved_at_utc': None, 'subreddit': 'depres...
15   t3  {'approved_at_utc': None, 'subreddit': 'depres...
16   t3  {'approved_at_utc': None, 'subreddit': 'depres...
17   t3  {'approved_at_utc': None, 'subreddit': 'depres...
18   t3  {'approved_at_utc': None, 'subreddit': 'depres...
19   t3  {'approved_at_utc': None, 'subreddit': 'depres...
20   t3  {'approved_at_utc': None, 'subreddit': 'depres...
21   t3  {'approved_at_utc': None, 'subreddit': 'depres...
22   t3  {'approved_at_utc': None, 'subreddit': 'depres...
23   t3  {'approved_at_utc': None, 'subreddit': 'depres...
24   t3  {'approved_at_utc': None, 'subreddit': 'depres...
25   t3  {'approved_at_utc': None, 'subreddit': 'depres...
26   t3  {'approved_at_utc': None, 'subreddit': 'depres...

In [ ]:
#LOOKS LIKE THIS DATA IS REALLY WHAT WE ARE LOOKING FOR
depress_json["data"]["children"][0]["data"]

{'all_awardings': [{'award_sub_type': 'GROUP',
   'award_type': 'global',
   'awardings_required_to_grant_benefits': 3,
   'coin_price': 300,
   'coin_reward': 250,
   'count': 1,
   'days_of_drip_extension': 0,
   'days_of_premium': 0,
   'description': 'THIS right here! Join together to give multiple This awards and see the award evolve in its display and shower benefits for the recipient. For every 3 This awards given to a post or comment, the author will get 250 coins.',
   'end_date': None,
   'giver_coin_reward': None,
   'icon_format': None,
   'icon_height': 2048,
   'icon_url': 'https://i.redd.it/award_images/t5_22cerq/vu6om0xnb7e41_This.png',
   'icon_width': 2048,
   'id': 'award_68ba1ee3-9baf-4252-be52-b808c1e8bdc4',
   'is_enabled': True,
   'is_new': False,
   'name': 'This',
   'penny_donate': None,
   'penny_price': None,
   'resized_icons': [{'height': 16,
     'url': 'https://preview.redd.it/award_images/t5_22cerq/vu6om0xnb7e41_This.png?width=16&amp;height=16&amp;auto

### 1.2 Creating functions to automate the Data Collection process 
- We will first run those functions on r/depression and check if they have worked well.

In [ ]:
# NOW WE CAN DEFINE A FUNCTION TO SCRAPE A REDDIT PAGE

def reddit_scrape(url_string, number_of_scrapes, output_list):
    #SCRAPED POSTS WILL BE CONTAINED IN OUTPUT LIST(SHD BE EMPTY)
    #THIS IS USEFUL FOR THE FIRST SCRAPE FROM THE VIRGIN SUBREDDIT
    after = None 
    for _ in range(number_of_scrapes):
        if _ == 0:
            print("SCRAPING {}\n--------------------------------------------------".format(url_string))
            print("<<<SCRAPING COMMENCED>>>") 
            print("Downloading Batch {} of {}...".format(1, number_of_scrapes))
        elif (_+1) % 5 ==0:
            print("Downloading Batch {} of {}...".format((_ + 1), number_of_scrapes))
        
        if after == None:
            params = {}
        else:
            #THIS WILL TELL THE SCRAPER TO GET THE NEXT SET AFTER REDDIT'S after CODE
            params = {"after": after}             
        res = requests.get(url_string, params=params, headers=headers)
        if res.status_code == 200:
            the_json = res.json()
            output_list.extend(the_json["data"]["children"])
            after = the_json["data"]["after"]
        else:
            print(res.status_code)
            break
        time.sleep(randint(1,6))
    
    print("<<<SCRAPING COMPLETED>>>")
    print("Number of posts downloaded: {}".format(len(output_list)))
    print("Number of unique posts: {}".format(len(set([p["data"]["name"] for p in output_list]))))
 

In [ ]:
#CALLING THE FUNCTION ON OUR DEPRESSION SUBREDDIT
depress_scraped = [] #DEFINING AN EMPTY LIST THAT WILL CONTAIN OUR SCRAPED DATA
reddit_scrape("https://www.reddit.com/r/depression.json", 50, depress_scraped)

SCRAPING https://www.reddit.com/r/depression.json
--------------------------------------------------
<<<SCRAPING COMMENCED>>>
<<<SCRAPING COMPLETED>>>
Number of posts downloaded: 1248
Number of unique posts: 997


In [ ]:
#CREATING A FUNCTION TO OUTPUT A LIST OF UNIQUE POSTS
def create_unique_list(original_scrape_list, new_list_name):
    data_name_list=[]
    for i in range(len(original_scrape_list)):
        if original_scrape_list[i]["data"]["name"] not in data_name_list:
            new_list_name.append(original_scrape_list[i]["data"])
            data_name_list.append(original_scrape_list[i]["data"]["name"])
    #CHECKING IF THE NEW LIST IS OF SAME LENGTH AS UNIQUE POSTS
    print("LIST NOW CONTAINS {} UNIQUE SCRAPED POSTS".format(len(new_list_name)))
    

In [ ]:
#CALLING THE FUNCTION ON OUR SCRAPED DATA
depress_scraped_unique = []
create_unique_list(depress_scraped, depress_scraped_unique)

LIST NOW CONTAINS 997 UNIQUE SCRAPED POSTS


In [ ]:
#PUTTING DEPRESSION DATA INTO A DATAFRAME AND SAVING TO CSV
depression = pd.DataFrame(depress_scraped_unique)
depression["is_suicide"] = 0
depression.head() #CHECK IF THERE ARE 100 COLS AND LAST DUMMY is_suicide COL

approved_at_utc   subreddit  ... author_cakeday is_suicide
0            None  depression  ...            NaN          0
1            None  depression  ...            NaN          0
2            None  depression  ...            NaN          0
3            None  depression  ...            NaN          0
4            None  depression  ...            NaN          0

[5 rows x 104 columns]

### 1.3 Running our functions on the r/SuicideWatch subreddit 

In [ ]:
#CALLING THE SCRAPING FUNCTION ON OUR SUICIDEWATCH SUBREDDIT
suicide_scraped = [] #DEFINING AN EMPTY LIST THAT WILL CONTAIN OUR SCRAPED DATA
reddit_scrape("https://www.reddit.com/r/SuicideWatch.json", 50, suicide_scraped)

SCRAPING https://www.reddit.com/r/SuicideWatch.json
--------------------------------------------------
<<<SCRAPING COMMENCED>>>
<<<SCRAPING COMPLETED>>>
Number of posts downloaded: 1230
Number of unique posts: 980


In [ ]:
#CALLING THE "UNIQUE ONLY" FUNCTION ON OUR SCRAPED DATA
suicide_scraped_unique = []
create_unique_list(suicide_scraped, suicide_scraped_unique)

LIST NOW CONTAINS 980 UNIQUE SCRAPED POSTS


In [ ]:
#PUTTING SUICIDEWATCH DATA INTO A DATAFRAME AND SAVING TO CSV
suicide_watch = pd.DataFrame(suicide_scraped_unique)
suicide_watch["is_suicide"] = 1
suicide_watch.head() #CHECK IF THERE ARE 100 COLS AND LAST DUMMY is_suicide COL

approved_at_utc     subreddit  ... author_cakeday is_suicide
0            None  SuicideWatch  ...            NaN          1
1            None  SuicideWatch  ...            NaN          1
2            None  SuicideWatch  ...            NaN          1
3            None  SuicideWatch  ...            NaN          1
4            None  SuicideWatch  ...            NaN          1

[5 rows x 104 columns]

#### NOTE: I've commented out the code in the next cell for "pd.to_csv" to prevent any accidental overwriting of the the saved dataset.**

In [ ]:
#suicide_watch.to_csv('../data/suicide_watch.csv', index = False)
#depression.to_csv('../data/depression.csv', index = False)

In [ ]:
#INVESTIGATING THE CASE OF r/SuicideWatch HAVING AN ADDITIONAL COLUMN
suicide_watch.columns.difference(depression.columns)

In [ ]:
#LOOKING INTO THAT ADDITIONAL COLUMN
suicide_watch['author_cakeday'].isnull().value_counts()

#### Early thoughts about the collected data
- Data seems to be collected successfully.
- We have some "uneven-ness" in the size of our set as we collected 980 r/SuicideWatch posts and 917 r/depression posts. We might want to consider "even-ing" out the posts with another round of collection. 
- There is also a matter of r/SuicideWatch having one extra column. Which is strange to me considering that they both exist on the same site. The column is "author_cakeday" and it is mostly NaNs. Thus, it doesn't seem like a column we will be using for our classifier.
